lib/awesome_spawn.rb

```diff
@@ -116,15 +138,94 @@ def build_command_line(command, params = nil)
   private
 
   def launch(env, command, spawn_options)
-    output, error, status = Open3.capture3(env, command, spawn_options)
+    if command.kind_of?(Array)
+      output, error, status = pipe_capture3(env, command, spawn_options)
+    else
+      output, error, status = capture3(env, command, spawn_options)
+    end
     return output, error, status && status.exitstatus
   end
 
-  def parse_command_options(command, options)
+  def normalize_run_opts(*opts)
+    commands, options = if opts.last.kind_of?(Hash)
+                          [opts[0..-2], opts.last]
+                        else
+                          [opts, {}]
+                        end
+
+    if commands.all? { |cmd| cmd.kind_of?(Array) }
+      commands = commands.flatten(1)
+    end
+
+    bad_keys = (options.keys.flatten & [:in, :out, :err]).map { |k| ":#{k}" }
+    raise ArgumentError, "options cannot contain #{bad_keys.join(", ")}" if bad_keys.any?
+
+    [Array(commands), options]
+  end
+
+  def parse_command_options(commands, options)
     options = options.dup
     params  = options.delete(:params)
     env = options.delete(:env) || {}
 
-    [env, build_command_line(command, params), options]
+    cmds = if commands.length == 1
```

Comment:
Instead of conditionals all over with special case handling, can we wrap single commands in an array and just run everything through that?

Reply:
Yeah, I couldn't think of a good way to do this, and all that I came up ended up being more complex than I was hoping.

Happy to take suggestions (can be pseudo code), but at the moment I don't have a better way to implement this off the top of my head, or specifics of things that you would like me to fix.  Still agree it is a bit kludgey under the covers.

Reply:
The problem with this suggestion is that it causes the specs to fail, with something like this:

```console
  12) AwesomeSpawn.run! result contains #command_line
      Failure/Error: expect(subject.send(run_method, "echo", :params => %w(x)).command_line).to eq("echo x")

        expected: "echo x"
             got: ["echo x"]

        (compared using ==)
      Shared Example Group: "executes" called from ./spec/awesome_spawn_spec.rb:368
      # ./spec/awesome_spawn_spec.rb:168:in `block (4 levels) in <top (required)>
```

The change from an `String` result to an `Array` means that when this is passed into `AwesomeSpawn#launch`, everything will run through `Open3.pipe_capture3`, which I don't want.  I preferably want (and I assume others as well) to only use that patch when a piped function is being passed, otherwise use the more battle tested and less involved `Open3.capture3` instead.